feat: rewrite and redesign

MarshallOfSound opened the pull request:
#### Description of Change

Brings our release website into 2025, all the same features but a brand new design and codebase that in theory makes it easy (and safe) to add new things. None of this handlebars nonsense anymore.

The new site is built entirely in React powered by [Remix](https://remix.run), all pages are server side rendered similar to the old handle bars site but transitions are client side, cached and way cleaner. It's themed using tailwind, the tailwind stuff I kinda understand now but I'll be honest I asked some LLM's for help with the design part (not my strong suite), turns out LLMs can spit out good tailwind based designs.

The new repo also maintains the old CI jobs, while adding a new build/lint/test one for the new codebase.

You can try out the new design / experience yourself right here on the [Staging Heroku Deployment](https://release-status-staging-7bcafebe7466.herokuapp.com/) that I spun up.  There are also a few screenshots below to give you an idea.

The review I'm looking for here:
* Things I missed / didn't convert from the old codebase
* Any bugs / issues with the new codebase
* If someone really prefers the old design I guess you can tell me 😆 
  * Note for all you light mode users out there, the screenshots are in dark mode but it does support your system theme, so light mode works too.

<details>
<summary>
Screenshots
</summary>


**Home**

<img width="1840" alt="image" src="https://github.com/user-attachments/assets/ba2ec0f9-7706-48df-9010-cd408434d423" />

**PR Lookup**

<img width="1840" alt="Screenshot 2025-04-27 at 8 59 16 AM" src="https://github.com/user-attachments/assets/e6309e3c-8712-4ec7-a581-ead3e00d31e9" />
<img width="1840" alt="Screenshot 2025-04-27 at 8 59 19 AM" src="https://github.com/user-attachments/assets/f79ed85d-5077-4b1f-ba46-1173734a4e2f" />

**Release**

<img width="1840" alt="Screenshot 2025-04-27 at 9 01 42 AM" src="https://github.com/user-attachments/assets/dc83666c-e38e-4289-88ae-9cae9143db35" />

**Release Comparison**

<img width="1840" alt="Screenshot 2025-04-27 at 9 02 26 AM" src="https://github.com/user-attachments/assets/71a74e5c-23be-478c-a55f-ec7730595606" />

**Release History**

<img width="1840" alt="Screenshot 2025-04-27 at 9 02 47 AM" src="https://github.com/user-attachments/assets/24865036-b9fb-423d-a929-62a8d4a218d7" />

</details>

dsanders11 commented on app/root.tsx:
```diff
@@ -0,0 +1,140 @@
+        <ArrowUpRight className="w-3 h-3" />
+      </>
+    ),
+    path: 'https://electronjs.org/docs',
```
```suggestion
    path: 'https://www.electronjs.org/docs/latest',
```

nit: nice to avoid the redirect

dsanders11 reviewed (changes requested):
> Things I missed / didn't convert from the old codebase

* The old site had links to the Chromium and Node.js releases when viewing an Electron release
* On the "All Releases" page, the ability to filter by major version was mildly useful
* When comparing versions, the "API Changes" view (e.g. https://releases.electronjs.org/release/compare/v34.3.3/v34.5.3/ts) was useful
* The history calendar view is less useful on the rewrite
  * The old version had a nightly release percentage for each month (e.g. "85.71% nightly")
  * The old version looked backwards on the past 12 months including the current month, which is more useful for seeing past releases. The new version shows the current calendar year and is not very useful in the first few months of the year.

> Any bugs / issues with the new codebase

* On the PR lookup page the input box has increment/decrement buttons which aren't necessary. Personal nit, it'd be nice if the input box could also take a URL and extract the PR number so you can just copy and paste without remembering to only grab the number.
* On the PR lookup details, the PR link for backports is `pr/null` and an empty "#" for branches awaiting manual backport
* On the PR lookup details, might be nice to set opacity to 50% to gray out the other semver types in "Semver Impact" just to make it more clear (or any other UI trick for showing they're unused).

Nuts and bolts:
* There's a few imports that include the `.js` suffix but the majority don't have it, so we should be consistent
* Tests are nice but inline snapshots are less nice. React Testing Library would be preferable (and what we're converting Fiddle to, slowly), but I'm not going to block this PR on it.

MarshallOfSound commented on .prettierrc.json:
```diff
@@ -3,5 +3,5 @@
   "tabWidth": 2,
   "singleQuote": true,
   "printWidth": 100,
-  "endOfLine": "lf"
```
Must have been a crazy line delete or something, wouldn't have done this deliberately lol

MarshallOfSound commented on app/routes/build/release-job.tsx:
```diff
@@ -0,0 +1,377 @@
+      return 'bg-green-100 text-green-800 dark:bg-green-900/30 dark:text-green-400';
+    case 'nightly':
+      return 'bg-purple-100 text-purple-800 dark:bg-purple-900/30 dark:text-purple-400';
+    default:
+      return 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-400';
```
Yeah type safety works when everything is in one repo, gotta handle bad data here

MarshallOfSound commented:
All good feedback, addressed and referenced below:

> The old site had links to the Chromium and Node.js releases when viewing an Electron release

Done

> On the "All Releases" page, the ability to filter by major version was mildly useful

It lives again

> The history calendar view is less useful on the rewrite

Nightly % is back, and you can now switch years which means that it's _more_ useful. I personally prefer views that show calendar years instead of rolling 12 months, this makes it the best of both worlds at the start of the year, you can just look at last years 👍 

> On the PR lookup page the input box has increment/decrement buttons which aren't necessary. Personal nit, it'd be nice if the input box could also take a URL and extract the PR number so you can just copy and paste without remembering to only grab the number.

URL regex implemented + type=tel instead of type=number to get rid of chromes silly buttons

> On the PR lookup details, the PR link for backports is pr/null and an empty "#" for branches awaiting manual backport

Oops, guess that wasn't a test case I was using, easy fix

> On the PR lookup details, might be nice to set opacity to 50% to gray out the other semver types in "Semver Impact" just to make it more clear (or any other UI trick for showing they're unused).

Did the opacity thing, should be more obvious now

> There's a few imports that include the .js suffix but the majority don't have it, so we should be consistent

Fixed, autocomplete I guess 🤷 

### Things not addressed yet

> Tests are nice but inline snapshots are less nice. React Testing Library would be preferable (and what we're converting Fiddle to, slowly), but I'm not going to block this PR on it.

The tests I added were mostly for self validation that I implemented things correctly / safely 😅 Inline snapshots make it super easy to review the file as input vs expectation instead of jumping around. If the snapshot is a few lines or less IMO inline make it easier to scan tests.

Remix has an RTL integration that makes it easy to test, just not super worried about that on the first pass, going from 0 tests to > 0 tests is a good start 😆 

> When comparing versions, the "API Changes" view (e.g. https://releases.electronjs.org/release/compare/v34.3.3/v34.5.3/ts) was useful

Lol forgot this was a thing, will figure out how to re-implement 🤔 Syntax highlighting + SSR sounds fun

#### Side note

Super funny to me that all the things you called out as missing / forgotton were things I implemented in the first place in the old one 🙃

MarshallOfSound commented on tsconfig.json:
Ejection of configs makes things more easy to trace IMO, the config is simple enough

MarshallOfSound commented on package-lock.json:
Lol, `create-remix-app` or whatever just decided npm was the way it would work, and tbh npm@10 isn't that bad 🤷 I can port it back to yarn quite easily, but maybe I should stop being as attached at some point.

_Character Growth_

erickzhao reviewed (commented):
Overall website LGTM. Great improvement from what we had before.

One thing is that we should make link colours stand out (via colour or underline) for accessibility purposes. Being familiar with the shape of our release notes, I can easily tell what's a link or not, but it's not very clear from just reading the rendered Markdown at the moment.

I think we can conform to WCAG standards around this by either adding colour contrast (with respect to regular text colour) or make links underlined by default.

A few pieces of feedback regarding the calendar:

1. Current date should not be marked as a "missed nightly". Maybe the overall calendar shouldn't be TZ-aware either, because we trigger the publish at specific times. For example, setting your time zone to Tokyo in DevTools will show tomorrow (4/30)'s nightly as missed even if we haven't even queued the scheduled 2025-04-30 job yet.

2. There's some weird discrepancy around the [02/03](https://release-status-staging-7bcafebe7466.herokuapp.com/history/2025-02-03), [02/04](https://release-status-staging-7bcafebe7466.herokuapp.com/history/2025-02-04), and [02/05](https://release-status-staging-7bcafebe7466.herokuapp.com/history/2025-02-05) nightlies where:
* 02/03 nightly shows in the box for 02/04
* 02/04 and 02/05 nightly are both bucketed into the box for 02/05

This doesn't appear to be the case in the original website.

MarshallOfSound commented:
(for the record fixed all the things in ericks comment)

MarshallOfSound commented:
Thanks for the feedback folks, per discussions all the things currently outstanding are non-blocking and to avoid this PR turning into a crazy 10 month thing, we're just going to land this as-is right now.

Then everyone can either raise PRs for their small things, or raise issues 👍
